formats/keep-json.ts

```diff
@@ -0,0 +1,199 @@
+                    this.importArchived = value;
+                });
+            });
+        
+		this.importTrashedSetting = new Setting(this.modal.contentEl)
+            .setName('Import deleted notes')
+			.setDesc('If imported, files deleted in Google Keep will be tagged as deleted. Deleted notes will only exist in your Google export if deleted recently.')
+            .addToggle(toggle => {
+                toggle.setValue(this.importTrashed)
+                toggle.onChange(async (value) => {
+                    this.importTrashed = value;
+                });
+            });
+
+		this.addOutputLocationSetting('Google Keep');
+
+	}
+
+	async import(progress: ProgressReporter): Promise<void> {
+		let { files } = this;
+
+		if (files.length === 0) {
+			new Notice('Please pick at least one file to import.');
+			return;
+		}
+
+		let folder = await this.getOutputFolder();
+		if (!folder) {
+			new Notice('Please select a location to import your files to.');
+			return;
+		}
+		let assetFolderPath = `${folder.path}/Assets`;
+
+		for (let file of files) {
+			try {
+				if(file.extension === 'zip') {
+					await this.readZipEntries(file, folder, assetFolderPath, progress);
+					
+				} else if(file.extension === 'json') {
+					let rawContent = await file.readText();
+					await this.importKeepNote(rawContent, folder, file.basename, progress);
+
+				} else {
+					const arrayBuffer = await file.read();
+					await this.copyFile(arrayBuffer, assetFolderPath, file.name, progress);
+				}
+
+			} catch (e) {
+				progress.reportFailed(file.name, e);
+			}
+		}
+
+	}
+
+	async readZipEntries(file: PickedFile, folder: TFolder, assetFolderPath: string, progress: ProgressReporter) {
+		await file.readZip(async zip => {
+			for (let entry of await zip.getEntries()) {
+				if (!entry || entry.directory || !entry.getData) return;
+				let curInnerFilename = '';
+				try {
+					let innerFileProps = parseFilePath(entry.filename);
+					curInnerFilename = innerFileProps.name;
+		
+					if(innerFileProps.extension === 'json') {
+						let rawContent = await entry.getData(new TextWriter());
+						await this.importKeepNote(rawContent, folder, innerFileProps.basename, progress);
+		
+					} else if(ATTACHMENT_EXTS.contains(innerFileProps.extension)) {
+						const rawContent = await entry.getData(new BlobWriter());
+						const arrayBuffer = await rawContent.arrayBuffer();
+						await this.copyFile(arrayBuffer, assetFolderPath, innerFileProps.name, progress);
+					}
+					// else: Silently skip any other unsupported files in the zip
+		
+				} catch(e) {
+					progress.reportFailed(`${file.name} / ${curInnerFilename}`, e);
+				}
+			}
+		})
+	}
+
+	async importKeepNote(rawContent: string, folder: TFolder, title: string, progress: ProgressReporter) {
+		let keepJson = convertStringToKeepJson(rawContent);
+		if(!keepJson) {
+			progress.reportFailed(`${title}.json (Invalid format)`, 'NOT A GOOGLE KEEP JSON');
```

Comment:
Can we update these to properly use a filename for the first parameter?

Reply:
Was going to flag this with you actually @lishid 
I'll change this, but please note that the ProgressReporter currently accepts a reason, but doesn't display it anywhere. Without that the user will have no idea why something was skipped, which will caused confusion.

Also, Are you happy with the format of the reasons specified in the second parameters?
I can't remember whether I put them in all caps for a reason.

Reply:
That's right. ProgressReporter is a work-in-progress that I have setup for the backend API - we are working in parallel on the interface to display the progress as well as the reason/details.

Everything will be user-visible in the end; I would prefer if they were just regular prose instead of allcaps.
Also the file path might end up getting parsed for better formatting of containing folder vs filename itself, so that's why it's important to use a proper filename instead of adding debug information into it.

Reply:
Makes sense, that was my expectation but just wanted to flag it.